### PR TITLE
test(core) #40: one harness for the vanilla MockConsumer tests, not five copies

### LINKED ISSUE #40: Reduce duplication in MockConsumer* test classes
## Context

The file-similarity tool on PR #34 flagged `MockConsumerTestWithCommitTimeoutException` and `MockConsumerTestWithEarlyClose` at 70.7% similarity (up from 65.1% after applying the same daemon-thread cleanup pattern to both).

More broadly, the test codebase has several standalone test classes (MockConsumerTest*, ProducerManagerTest) that duplicate infrastructure that could be shared.

## Scope

1. **MockConsumer* tests**: Extract common setup (mock consumer creation, record-adder daemon thread, PC lifecycle, cleanup) into a shared base class or helper. Each test overrides only the failure behaviour and assertion.

2. **General test deduplication audit**: Review the file-similarity report for other high-similarity pairs in `src/test/` and `src/test-integration/` and refactor where the duplication is structural (not just "similar test shape").

## Not blocking

The 70.7% is within the `fail_above: 80` threshold. This is cleanup work, not urgent.

Surfaced during PR #34 CI reliability work.

### PR DESCRIPTION
**Pull request by astubbs**
Closes astubbs#40 - the first half by extraction, the second by a documented verdict (see *Secondary audit* below).

## Description

The six `MockConsumer*` test classes each carried their own copy of the same wiring: build a `MockConsumer`, hand-rebalance the partition in, tell PC about it separately, update beginning offsets, construct and subscribe PC, feed records, collect them out of the user function, tear down. The similarity check flagging two of them at 70.7% on astubbs#34 is the symptom; the cause is that the wiring was copied rather than shared, so it drifted - the daemon-thread comment about PIT attributing a stray `addRecord()` to whatever test runs next in the JVM appears verbatim in two files, and its cleanup exists only in those two.

**`MockConsumerTestBase` now owns that wiring.** A scenario supplies its failure behaviour (`createMockConsumer()`) and the options it needs (`customiseOptions(..)`), and nothing else.

- It deliberately does **not** extend `AbstractParallelEoSStreamProcessorTestBase`. That base wires a Mockito-spied `LongPollingMockConsumer` and a `MockProducer`; the whole subject of these tests is what PC does when the consumer misbehaves in ways only a hand-written `MockConsumer` subclass can express. The javadoc says so, so a future session does not "fix" it.
- `CommitRejectionTestBase`, which had already extracted its own half of the same wiring, now sits on the harness too - so a third rejection reason is one method rather than another copy.

**Deliberately left duplicated:** every scenario keeps its own Awaitility block, with its own timeout, in its own file. Those are the point of each test, and the timeouts are scenario-specific - each has to clear that scenario's simulated outage window. Hoisting them would mean opening two files to learn what a test actually checks, which is a net loss even though it would cut more lines.

**Teardown is now uniform and strictly stronger than what it replaced:** the record feed is interrupted *and joined* (it was interrupted only, and only in two of the classes), before PC is closed rather than after, and `Awaitility.reset()` runs first so a throwing close cannot skip it.

Teardown closes PC with `close()`, which is the **non-draining** close (`DrainingCloseable.close()` delegates to `closeDontDrainFirst()`). Review flagged that `CommitRejectionTestBase` was overriding a `closeParallelConsumer()` hook to call `closeDontDrainFirst()` explicitly - functionally identical to the default, so the hook had no real user and its javadoc ("override where draining first is wrong") stated the opposite of what the default did. Hook and override both removed; the fact is now a comment at the one place that closes.

### Two real defects found on the way

- **`@Timeout(60000L)` on three of these classes meant 60000 _seconds_** - JUnit's default unit - i.e. no timeout at all. Replaced with `@Timeout(120)` on the base (`@Timeout` is `@Inherited`), real headroom over the longest scenario (25s measured) and an actual guard against a wedged `MockConsumer` test.
- **`MockConsumerRebalanceInProgressTest`'s javadoc still linked `MockConsumerTestWith{CommitTimeout,SaslAuthentication}Exception`**, renamed away in dc44e202. `javac` does not check `{@link}` without doclint, so it rotted silently.

### One assertion added, none weakened

`MockConsumerEarlyCloseTest` asserted only that `close()` returned. It now also asserts PC ended closed **with no failure cause** - "shut down cleanly", as its javadoc has always claimed, rather than "died in a way that also reports closed". Verified to hold.

### Secondary audit (astubbs#40's second half): done, deliberately not acted on

The remaining high-similarity pairs across `src/test/` and `src/test-integration/` are overwhelmingly **cross-module clones** - `MutinyBatchTest`/`ReactorBatchTest` in the **high 70s**, `MutinyPCTest`/`ReactorPCTest` in the **low 70s**. Deduplicating those means a generified test base in core's test-jar that each module parameterises with its own processor type: a different and much larger job, and exactly the scope creep astubbs#40 says it does not want. The largest *within-module* test pairs are `VeryLargeMessageVolumeTest` against `MultiInstanceHighVolumeTest` (**high 50s**) and `TransactionAndCommitModeTest` (**mid 50s**) - both well under the check's `fail_above: 80`, and broker ITs that want Docker to verify, not a desk refactor. The `TestConventionsArchTest` x4 at **89-91%** are already documented as irreducible.

> **Correction.** An earlier revision of this section - and of the `docs/refactoring.md` table it summarises - carried *estimated* figures (~94 / ~86 / ~80 / ~88 / ~98%). Review caught it, and re-measuring showed every one was wrong, by 7 to 48 points. Two were wrong *claims* rather than stale decimals: `TransactionAndCommitModeTest`/`VeryLargeMessageVolumeTest` was called "the only within-module pair above `fail_above: 80`" when it measures **53.9** and ranks sixth, and `MutinyUnitTestBase`/`ReactorUnitTestBase` was listed as part of "the real prize" at ~80% when it measures **32.0**, barely over the `ignore_below: 30` floor. Both are now measured off the `duplicate-code-detection-tool` report this PR's own CI posts, and quoted as bands so they do not drift - see 3994b6c0.

All of it is ranked with a verdict in `docs/refactoring.md` so the next reader does not re-derive the audit. `ProducerManagerTest`, named in the issue, turned out to have no cross-file duplication left: it already uses the shared `PCModuleTestEnv`/`ModelUtils` infrastructure, and what it repeats is one line of `Awaitility.reset()`.

**Similarity check: the prediction was wrong, and the numbers are in.** This PR expected the
extraction to push `CommitRejectionTestBase` ↔ `MockConsumerTestBase` to ~70% - two small abstract
harnesses in one package sharing imports. It did not happen: **neither file appears in the report at
all**, against anything, and the check's reporting floor is 30%. What the run actually shows is the
thing astubbs#40 asked for - the `MockConsumer*Test` scenarios now pair in the **34-37%** band
(`CommitTimeout`↔`EarlyClose` and `CommitTimeout`↔`Sasl` around 37%, `EarlyClose`↔`Sasl` around
34.5%), down from the **70.7%** that astubbs#34 flagged. The measure is corpus-relative, so the
decimals drift with unrelated changes - this pair moved 37.49 -> 37.43 across the master merge, which
touched none of these files; the band is the durable part. PMD CPD: no new clones, duplication -0.39%. jscpd: -0.42%, and its one "new clone" is the
8-line package + copyright + import block shared by `MockConsumerEarlyCloseTest` and
`MockConsumerSaslAuthenticationTest`, every line of which both files use. Docs corrected to the
measured values rather than the estimate.

### Notes

- `MockConsumerTestBase` is an extraction from upstream-derived files, so it keeps the Confluent header plus the modifications line and is registered in `EXTRACTED_FROM_UPSTREAM` in `bin/check-copyright-headers.sh`.
- No product code changed - test sources, one scanner registration, and docs.

### Verification

- The six tests pass, **three consecutive runs, no intermittency** (nothing became flaky).
- Full unit suite (`bin/ci-unit-test.sh`, all modules): green.
- `bin/check-copyright-headers.sh` and its self-test: green.

## Checklist

- [x] Docs updated - `docs/refactoring.md` (the deferred cross-module clone backlog, with measured figures) and `docs/test-hardening/inactive-tests-audit-2026-08-08.md` (re-pointed at this PR's renames; its 9.4 finding deleted, now resolved on both counts). An earlier revision also added `docs/inflight/test-mockconsumer-harness.md`; that has been **deleted** - review pointed out it did not earn a file, since git announces the conflicts it warned about and the javadoc already carries the only part no command supplies.
- [x] Tests added/updated - this PR is entirely test work; the six tests test the same things, plus one strengthened assertion in `MockConsumerEarlyCloseTest`
- [x] Title & body reflect the final content of this PR
- [x] Self-hosted runner / security implications considered - `N/A` - no workflow or runner changes; the only non-test file is a path registration in `bin/check-copyright-headers.sh`

### CODE REVIEW

**Comment by astubbs** [comment]
@claude review this

Context for the reviewer: the scheduled `claude-review` run on this head SHA ([run 30979310097](https://github.com/astubbs/parallel-consumer/actions/runs/30979310097)) completed with `is_error: false` after 52 turns but never wrote its findings into its comment - it left two todo items unchecked and posted no inline comments. So this PR has not actually been reviewed yet. Please do a full review pass.

Particular things worth your attention:
- The teardown ordering in the new `MockConsumerTestBase` (`Awaitility.reset()`, then interrupt **and join** the feed thread, then close PC) - is that ordering actually safe for all six scenarios?
- `@Timeout(120)` on the base relies on `@Timeout` being `@Inherited`. Please verify that inheritance actually applies to the subclasses.
- `MockConsumerEarlyCloseTest` gained an assertion (`getFailureCause()` is null). Is that assertion actually meaningful, or can it pass vacuously?

**Comment by astubbs** [comment]
Thanks - that's the first substantive review this PR has had, and it found something real. Point-by-point:

### The one finding I acted on: the redundant close hook

You flagged `CommitRejectionTestBase.closeParallelConsumer()` as functionally identical to the base default and said you'd leave it as clarifying. **I went further and removed it**, because I think it was mildly misleading rather than clarifying, and the base javadoc was outright wrong.

Verified your claim first - `DrainingCloseable.close()` is `{ closeDontDrainFirst(); }` (`DrainingCloseable.java:32-33`), and `AbstractParallelEoSStreamProcessor.close()` overrides it with the same delegation (`:568-571`). So the override changed nothing.

The problem is what the hook's javadoc said:

> Override for a scenario where draining first is wrong

That asserts the default **drains**. It does not. A future author adding a seventh scenario and trusting that sentence would reason backwards about teardown - and the one override present appeared to confirm it, since "overrides to not drain" only makes sense if the default drains. The pair was self-reinforcingly wrong.

Since the hook had no override that changed behaviour, it was also speculative generality. Both are gone; the non-draining contract and *why teardown wants it* (it runs on the failure path too, so it must not hang waiting on in-flight work against a consumer still being made to misbehave) is now stated once, at the only place that closes. A scenario that ever wants the opposite should call `closeDrainFirst()` in its own test body, where the wait is visible. Fixed in 2d1b2a47.

### Your three answers - agreed, and independently confirmed

- **`@Timeout` inheritance.** Confirmed from bytecode rather than reasoning: `javap -v org/junit/jupiter/api/Timeout.class` shows `java.lang.annotation.Inherited` on the annotation type, and `unit()` has `AnnotationDefault: TimeUnit.SECONDS` - which is also the proof that the old `@Timeout(60000L)` meant 60000 seconds (~16.7 hours), i.e. no timeout.
- **Teardown ordering.** Agreed. Worth adding one case you didn't mention: `interrupt()` cannot free a thread blocked *entering* a `synchronized` monitor, and `MockConsumerCommitTimeoutTest`'s mock holds the `MockConsumer` monitor for 5s per failing `commitSync`. So the feed thread can genuinely be unfreeable for a few seconds. That is why the join is bounded and logs rather than hangs - worst case teardown costs 10s, well inside the 120s budget. The outage is also over by the time that test tears down.
- **`getFailureCause()` assertion.** Agreed, non-vacuous, same trace.

### On the test evidence you couldn't gather

You flagged that you couldn't run `./mvnw` - noted, and you were right to say so rather than imply verification. I ran the six locally: **6/6 pass**, `MockConsumerTest` 0.4s, `RebalanceInProgress` 1.1s, `CommitFailed` 1.1s, `EarlyClose` 5.3s, `Sasl` 10.4s, `CommitTimeout` 25.3s. That 25.3s worst case is what sizes `@Timeout(120)`. Re-run green after the hook removal above.

(`-pl parallel-consumer-core` alone trips `enforcer:enforce`'s `ReactorModuleConvergence`; add `-Denforcer.skip=true` or `-am`.)

### The similarity prediction in the description was wrong, and is now corrected

The description told you to expect `CommitRejectionTestBase` <-> `MockConsumerTestBase` at ~70%. **That did not happen** - neither file appears in the report at all, against anything, and the floor is 30%. The prediction came from a stdlib stand-in I wrote because the real tool won't build locally (`gensim` fails on py3.14; on py3.11 an `nltk` import shim blocks imports).

What the run actually shows is the result #40 wanted: the `MockConsumer*Test` scenarios pair at **34-37%**, down from the **70.7%** that #34 flagged. PMD CPD: no new clones, -0.39%. jscpd: -0.42%, and its single "new clone" is the 8-line package + copyright + import block shared by `MockConsumerEarlyCloseTest` and `MockConsumerSaslAuthenticationTest` - every line used by both, not actionable in Java.

Description and both docs corrected to the measured values in 762087f2, including the lesson (read the tool, don't model it).

### Still standing, not changed

Not extending `AbstractParallelEoSStreamProcessorTestBase` (it wires a Mockito-spied `LongPollingMockConsumer` + `MockProducer`; these tests exist to exercise behaviour only a hand-written raw `MockConsumer` can express), and leaving each Awaitility block in its own subclass with its own scenario-specific timeout. You didn't challenge either - recording that they were deliberate, not overlooked.

**Comment by astubbs** [comment]
## The red `claude-review` check is a bug in `bin/check-review-posted.sh`, not a finding on this PR

Reporting, **not fixing** - it's a CI gate, it's on `master` (added in a4db924a), and it affects every PR, so it isn't mine to patch from a feature branch. Leaving the check red.

### Diagnosis

`bin/check-review-posted.sh:94` is:

```bash
if printf '%s\n' "$comment_bodies" | grep -qE "actions/runs/${run_id}([^0-9]|$)"; then
```

under `set -euo pipefail` (line 56). `grep -q` **exits the moment it matches**. If `printf` still has data to write, it takes `SIGPIPE`/`EPIPE` and exits non-zero. `pipefail` then makes the whole pipeline non-zero, the `if` takes the false branch, and the script reports "no comment cites this run" - *even though grep matched*.

The failing run left the proof in the log, immediately above its own error:

```
bin/check-review-posted.sh: line 94: printf: write error: Broken pipe
##[error]The review step reported success but posted no review: no comment ... cites run 30979310097
```

A broken pipe on line 94 can only happen **because grep matched and exited early**. The gate's evidence of success is what makes it fail.

### Reproduced three ways against this PR's real comment stream

| Input | Citation position | Result |
|---|---|---|
| 37 bytes | present | `Review posted by run 30979310097.` exit 0 |
| 130,807 bytes (real comments) | comment #2 of 5 | **exit 1 - false failure** |
| ~100KB synthetic | last line | exit 0 |

Same needle, same script - only the position and the size change. It fails when the citation is found early in a stream larger than the 64KB pipe buffer, and passes when grep is forced to read to the end.

### Why it triggered here

The similarity report comment is **127,686 bytes** of the 130,807 total. The `claude[bot]` comment carrying the citation is #2; the giant report is #4. So grep matches ~1KB in and exits with ~128KB still unwritten - guaranteed EPIPE. Any PR with a large similarity report hits this.

### Two separate things, both worth knowing

1. **The gate is wrong here.** Run 30979310097 *did* post a comment citing itself, so a working gate would have gone green.
2. **But the review was substantively empty anyway.** That run finished `"is_error": false`, 52 turns, $2.27, 14 permission denials - and never wrote its findings into its comment, leaving two todo items unchecked and posting zero inline comments. That's exactly the "posts its progress comment and then produces an empty review would still pass" limit the script's own header calls out. Worth knowing that the failure mode is real and not hypothetical.

I got a real review by asking for one in a comment (#issuecomment-5188876914), as the script's error message suggests - it ran clean in 6m27s and found one genuine issue, now fixed.

### Note on the fix, for whoever picks it up

`bin/test-check-copyright-headers.sh` already asserts `ok: scanner has no SIGPIPE-prone pipes into grep -q / awk` - the repo has met this bug class before and guards `check-copyright-headers.sh` against it. `check-review-posted.sh` shipped in the same commit without that guard, and has no self-test. The obvious candidates are dropping `-q` (`grep -cE ... > /dev/null`), a here-string (`grep -qE ... <<< "$comment_bodies"`), or a `case`/`[[ ]]` match with no pipe at all - plus a large-input regression test, since a small-input test passes today.

---

### Update: the gate went green on the next push, and that *confirms* the diagnosis rather than clearing it

`claude-review` passed on run 30986389457. Nothing was fixed - the bug is still there, and the reason it passed is the smoking gun.

The gate greps for the run id in the comment stream, which GitHub returns in creation order. Compare where the citing comment sat:

| Run | Position of the comment citing that run | 127,696-byte similarity report | Result |
|---|---|---|---|
| 30979310097 | **2nd of 5** | *after* the match | grep exits early -> `printf` EPIPE -> **false fail** |
| 30986389457 | **10th of 11** | *before* the match | grep must read to the end -> no EPIPE -> **pass** |

Same script, same needle, same repo - only the match's position relative to the 64KB pipe buffer changed. A gate whose verdict depends on whether the thing it is looking for happens to appear before or after an unrelated bot's large comment is not measuring what it thinks it is.

So this is **latent, not resolved**. It will fail again on any PR where the reviewer's comment lands before a large one - which is the normal case early in a PR's life, before the reviewer has re-run. Worth fixing on `master` with a large-input regression test, since a small-input test passes today.

**Comment by astubbs** [comment]
## Chaos Pain Suite is red, and it is **not** this PR - it reproduced the open silent-stall bug

Diagnosed rather than retried-to-green. Recording the evidence so nobody has to re-derive it.

### What actually failed

`ChaosChurnStormIT.churnStormMeetsSlosAndBalancesLedger` - not a `MockConsumer` test. From the failsafe report:

```
Run summary: consumed=100387, probe violations=[NO_PROGRESS: fleet consumed
count stuck at 98800/100000 for 30s (bound 30s)]

violations (23):
  - CLASS2_STALL/LAG_STAGNATION: partition ...-34 lag=625 with committed offset
    stagnant at 653 for 154s (bound 150s) - protocol-invisible stall: group STABLE
    + heartbeats flowing, yet this partition's backlog is going nowhere
  - CLASS2_STALL/LAG_STAGNATION: partition ...-21 lag=1132 ... stagnant at 91 for 154s
  - CLASS2_STALL/LAG_STAGNATION: partition ...-33 lag=1128 ... stagnant at 47 for 154s
  ... (23 total)
```

That is the **silent-stall-after-rebalance** class (#857) - group STABLE, heartbeats flowing, backlog frozen. Known, root cause still open. The chaos suite did its job; this is a real product signal, not noise, and it should **not** be quarantined or bounded away.

### Why it cannot be this PR

- **Zero main-code changes**: `git diff a4db924a..HEAD -- '*/src/main/*'` is empty. Nothing this PR touches can execute in a broker run.
- **No linkage**: nothing in `src/test-integration/` references `MockConsumerTestBase`, `CommitRejectionTestBase`, or any changed test class. `ChaosChurnStormIT` extends `ChaosScenarioBase`, an unrelated hierarchy.
- **Same code passed already**: this suite was **green on this PR's previous head** (run 30979310092, 6m26s). Between the two runs the only Java change was deleting a no-op override.
- **Ambient right now**: `fix/release-0600-blockers` failed the same suite at 07:29:37 in the same window, while other branches passed.

### Runner saturation is the aggravating factor, not the cause

The failing run logged the load signature throughout - `Thread execution pool termination await timeout (PT10S)`, `Clean execution pool termination failed ... Threads still not done count: 3`, and repeated `records in the queue have been waiting longer than 10s`. Ten branches were hammering the self-hosted runners concurrently. Note that the stall bound that tripped is 154s against a 150s bound - a 3% overshoot, exactly what you'd expect a real-but-marginal stall to look like when CPU-starved.

That is context for *why it surfaced now*, **not** a reason to raise the bound. The probe is measuring the right thing.

### What I did and did not do

- Did **not** touch the probe, the bounds, the quarantine registry, or any gate.
- Did trigger one re-run of the failed job as a **diagnostic** (deterministic vs intermittent), not to buy a green tick. At the time of writing it has been queued ~25 minutes behind the saturated self-hosted runner pool and has not started.

Whatever it reports, the analysis above stands: this suite's failure is independent of a test-only refactor with no main-code delta. If it goes green on re-run, that confirms intermittency under load; if it fails again, that is a stronger #857 datapoint and worth attaching to that issue rather than to this PR.

**Comment by astubbs** [comment]
## Scope note for whoever picks this up

**This PR now closes #40** - the description was changed to carry a real closing link, so #40 will close on merge. Flagging it because the original wording said "closes the first half", and the second half is closed by *verdict* rather than by code.

**First half - done by doing it:** `MockConsumerTestBase` extracts the shared wiring; the six scenario classes now supply only their failure injection and assertions. Measured effect: the `MockConsumer*Test` scenarios moved from the 70.7% that triggered #40 down to **34-37%**.

**Second half - done by documented decision, not by code.** The audit ran. The remaining high-similarity pairs are overwhelmingly cross-module clones (`MutinyBatchTest`/`ReactorBatchTest` ~94%, `MutinyPCTest`/`ReactorPCTest` ~86%), and deduplicating those needs a generified base in core's test-jar spanning three modules - the exact scope creep #40 says it does not want. The one within-module pair over `fail_above: 80` is broker ITs needing Docker to verify. `TestConventionsArchTest` x4 is documented-irreducible. `ProducerManagerTest`, named in #40 itself, already has no cross-file duplication left. All ranked with verdicts in `docs/refactoring.md`.

**If you disagree that a documented won't-do discharges the second half, remove the `Closes #40` line before merging** and open a follow-up issue for the cross-module work. Leaving #40 open without that follow-up is the worst option - the verdict then gets re-derived from scratch by whoever picks it up next.

**One caveat on the duplication numbers:** the real similarity tool could not be run locally (`gensim` will not build on py3.14; on py3.11 an `nltk` import shim blocks imports), so the pre-merge prediction came from a stdlib stand-in and was wrong - it predicted the two abstract harnesses would register ~70%, and neither appears in the CI report at all. The 34-37% figures above are from the authoritative CI run, not the stand-in.

**Comment by astubbs** [comment]
> Worth either correcting the five figures to match the linked CI report, or loosening them to a band the way the `MockConsumer*Test` figures were revised to (`34-37%` rather than fixed decimals), so they don't need re-verification on every future merge either.

Fixed in 3994b6c0 — took both halves of that: measured against the report, then quoted as bands.

You were right that it was pointed, and re-measuring turned up more than stale decimals. Two of the five were **wrong claims**, not just wrong numbers:

- **`TransactionAndCommitModeTest` ↔ `VeryLargeMessageVolumeTest`** was described as *"the only within-module pair above the check's `fail_above: 80`"*. At **53.9** it is not above 80 — and it is not even the largest within-module pair. It is sixth, behind two production-code pairs (`RunLengthV1`/`V2EncodingNotSupported` ~63, `JStreamParallelEoSStreamProcessor`/`JStreamParallelStreamProcessor` ~62) and behind `MultiInstanceHighVolumeTest`/`VeryLargeMessageVolumeTest` at **57**, which the section did not mention at all. Rewrote the entry around `VeryLargeMessageVolumeTest`, since that is the actual shared shape (one volume-test reused three ways — a single extraction, not two), and named the production pairs as out of scope so the next reader doesn't rediscover them and assume they were missed.
- **`MutinyUnitTestBase` ↔ `ReactorUnitTestBase`** was listed as part of *"the real prize"* at ~80. At **32.0** it barely clears the `ignore_below: 30` floor. The bases already share core's test-jar; the duplication is in the scripts. Listing it was inviting someone to extract what is already extracted.

Corrected figures, all measured off the report at this head:

| Pair | Doc said | Measured |
|---|--:|--:|
| `MutinyBatchTest` ↔ `ReactorBatchTest` | ~94% | **78.9** |
| `MutinyPCTest` ↔ `ReactorPCTest` | ~86% | **71.1** |
| `MutinyUnitTestBase` ↔ `ReactorUnitTestBase` | ~80% | **32.0** |
| `TransactionAndCommitModeTest` ↔ `VeryLargeMessageVolumeTest` | ~88% | **53.9** |
| `TestConventionsArchTest` (6 pairings) | ~98% | **89.5–90.9** |

One correction to your table: `MutinyPCTest`/`ReactorPCTest` and the others moved slightly between the run you read and this head (71.18 → 71.09, 31.93 → 32.02, 60.73 → 53.94) — which is the drift argument for bands, so the section now quotes bands rather than decimals, matching what ba77e67c did to the `MockConsumer*Test` figures.

Also recorded why the `TestConventionsArchTest` pairs sit over `fail_above: 80` without failing the build: the check runs `compare_with_base: true`, and they appear in neither the report's *new* nor its *increased* section.

Separately, this branch is now merged up from master (which had landed its own `io.confluent.*` → `bz.stub.*` rename, so the branch's copy of it collapsed out). `MockConsumerEarlyCloseTest` was the one real content collision — master's #265 replaced its `Thread.sleep(5000)` with an Awaitility wait on two *poll-path* auth failures, and that now lives inside the harness rather than either side winning. `git diff origin/master...HEAD` is back to exactly the 10-file payload. Full unit suite green: 411 tests, 11 modules, 0 failures.

**Comment by astubbs** [comment]
### Babysit run report — chaos lane RED, parked as unrelated

`highcpu / Chaos Pain Suite` went red on `3994b6c0`. **Not caused by this PR, and not blocking it** (non-required — merge state is `UNSTABLE`, not `BLOCKED`). Recording it here rather than in the PR body because it belongs to the Class 2 hunt, not to #40.

**What fired.** `ChaosRevokeUnderWorkIT.revokeUnderWorkStaysProtocolHonest` — 15 `CLASS2_STALL/LAG_STAGNATION` violations, **13 of 80 partitions with committed offset frozen for 154s against the 150s bound**, group `STABLE`, heartbeats flowing. Cooperative arm and `ChaosChurnStormIT` both passed. Seed `3341129570805008712`, [run 31668991207](https://github.com/astubbs/parallel-consumer/actions/runs/31668991207/job/94349595121).

**Why it is not this PR.** The payload is 6 MockConsumer *unit* tests, 2 docs files and one line in `bin/check-copyright-headers.sh` — no production code, nothing the chaos ITs touch. The "green on every other branch, red only here" signal dissolves on inspection: `bin/chaos-test.sh:20` uses a **random seed** unless `CHAOS_SEED` is set, so each run is a different schedule, not a different codebase. I also considered master's #265 (which moved `junit-platform.properties` into surefire/failsafe `configurationParameters` and could have changed IT parallelism) and falsified it — three branches carrying #265 passed this lane.

**Why it may be real rather than the known thin-margin flake.**

| Evidence | Value |
|---|---|
| Frozen partitions' combined lag | **34,686** |
| Run shortfall (250,000 − 213,306) | **36,694** — the frozen partitions are **94.5%** of it |
| Fleet progress during quiet phase | 172,104 → 213,306 (rest of the fleet drained) |
| All 13 froze at | the same instant, 154s |
| Modelled worst *legit* freeze | **100s** (60s storm + 20s heavy + ~20s slack, `AbstractRevokeUnderWorkScenario.java:52-56`) |
| Previously measured legit peaks | **95-112s** (9-seed master sweep), **117-123s** (W4) |

So the remaining backlog was concentrated in the stalled partitions while everything else finished — that is not a loaded runner, and 154s is above every legit peak measured to date.

**What I deliberately did not do.** [`docs/inflight/test-chaos-phase2.md`](../blob/master/docs/inflight/test-chaos-phase2.md) pre-authorises widening — *"Thin margin… widen it if it ever flakes"* — but the same ledger says *"the RED side awaits a real occurrence"*. Widening `LAG_STAGNATION_BOUND` on the exact run where the tripwire may finally have caught its quarry is the worst moment to blunt it, and it is a calibration call, not a defect repair. It also has no business landing in a test-dedup PR.

**Suggested next step (one command, decides it):**

```
CHAOS_SEED=3341129570805008712 bin/chaos-test.sh
```

If it reproduces, this is the Class 2 occurrence the hunt was built for. If not, it goes in the ledger as the first legit peak over 150s and the thin-margin line becomes evidence-based.

I have not established *why* those 13 partitions stopped — that needs the replay — so no fix is proposed here.

**Comment by astubbs** [comment]
### Chaos RED resolved: seed replay says thin-margin flake, not a Class 2 stall

Replayed the failing schedule. **My earlier lean in the run report above was wrong** — this is the flake `docs/inflight/test-chaos-phase2.md` predicted, not the real occurrence the RED-side hunt is waiting for.

**Same seed, same schedule, two environments:**

| | `maxLagStagnation` | Outcome |
|---|--:|---|
| CI (`31668991207`, self-hosted highcpu) | **154,460ms** | RED — 4.5s over the 150s bound |
| Local replay (uncontended 32-thread box) | **121,264ms** | GREEN — 0 violations, drained **252,665** ≥ 250,000 |

`CHAOS_SEED=3341129570805008712 bin/chaos-test.sh` → `BUILD SUCCESS`, all three scenarios green, `ChaosRevokeUnderWorkIT` 159.4s.

**Why this settles it.** A real Class 2 stall is *unbounded* — that is the probe's whole premise. On replay the identical schedule drained past the expected count with zero violations, so the schedule does not encode a wedge. The 121s peak also lands squarely in the ledger's already-measured legit band (117-123s). What differs is not the schedule but the clock.

**Why CI's clock was slower — the box was shared.** Both `Performance` and `Chaos Pain Suite` are matrix jobs on `runs-on: [self-hosted, highcpu]`, and on this run they overlapped:

- Performance: `05:03:41` → `05:06:57`
- Chaos: `05:03:57` → `05:11:28`

The overlap covers the chaos storm phase and the stagnation measurement window. The workflow header already says this out loud: *"its timing is noisy when it shares the box with chaos (accurate benchmarking belongs in a separate isolated/on-demand run)"*. That contention stretched the legitimate commit-freeze window by ~27% (121s → 154s) and pushed it over a bound whose modelled worst legit case is only 100s.

So the ledger's *"Thin margin… widen it if it ever flakes"* line is now evidence-backed rather than a guess: the legit distribution demonstrably reaches 154s on a shared runner, and the bound sits at 150s.

**Still not fixing it here.** Two candidate remedies — widen `LAG_STAGNATION_BOUND` / shorten the storm, or stop Performance and Chaos sharing the box — are both chaos-lane calibration work with no business in a MockConsumer test-dedup PR. Recording the measurement so whoever picks it up starts from data.

Net effect on this PR: **none.** The lane is non-required, the failure is environmental, and nothing in this PR's 10 files can influence it.

File: docs/inflight/test-mockconsumer-harness.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
+# The vanilla-`MockConsumer` test harness
```
Comment (astubbs) [inline docs/inflight/test-mockconsumer-harness.md]: needs update

File: bin/check-copyright-headers.sh [hunk 2]
```diff
@@ -152,6 +152,7 @@ ${COPYRIGHT_CHECK_EXTRA_RENAMES:-}
 # checked) and stops the check ambushing the stack at merge (see PR astubbs#83).
 EXTRACTED_FROM_UPSTREAM="
 parallel-consumer-core/src/test-integration/java/bz/stub/parallelconsumer/integrationTests/utils/ManagedPCInstance.java
+parallel-consumer-core/src/test/java/bz/stub/parallelconsumer/MockConsumerTestBase.java
```
Comment (astubbs) [inline bin/check-copyright-headers.sh:155]: now that we've done the full package rename, does this extracted from upstream system still need to be there?

**Review by astubbs — COMMENTED** [review]
minor

File: bin/check-copyright-headers.sh (hunk 2 above)
Comment (astubbs) [reply bin/check-copyright-headers.sh:155]: Yes — provenance is about who wrote the code, not what package it sits in, so the rename doesn't touch it: deleting this one line takes `bin/check-copyright-headers.sh` from 0 violations to 1 (the file loses its fork-point origin, is judged fork-original, and the Confluent header it correctly carries becomes the violation).

File: docs/inflight/test-mockconsumer-harness.md (hunk 1 above)
Comment (astubbs) [reply docs/inflight/test-mockconsumer-harness.md]: > needs update

Agreed, and done in dcc3d9fa — though for a bigger reason than staleness, so worth saying what changed.

Measured against `docs/inflight/AGENTS.md`'s first rule ("Track only what is currently OPEN"), only one of the file's three sections belonged there at all:

| Section | Verdict |
|---|---|
| "There is now one harness…" | Settled design rationale — already in `MockConsumerTestBase`'s class javadoc |
| "Collides with" | **Genuinely in flight** — astubbs#202 is open, and every branch forked before this conflicts on these six files |
| "Reading a file-similarity comment" | Duplicated `docs/refactoring.md`, which the file itself linked to while restating it in full |

The harness isn't in flight — it lands with this PR. So the note is now just the transient part: the wiring moved out of six files, so a branch that forked earlier conflicts, and the resolution is "keep the harness, re-apply your scenario" rather than the reverse (re-applying a pre-harness copy silently un-does the extraction). Plus the astubbs#202 pointer. It also states its own deletion condition now.

I checked each removed line survives in its permanent home before deleting it — the javadoc for the rationale, `docs/refactoring.md` for the figures.

Two follow-ons from this, both already pushed:
- `dce042b29` — `docs/inflight/AGENTS.md` was correctly written and correctly linked, and still got missed. It only appeared in the *ownership* table, not the "read this when doing X" routing table, so it read as reference rather than as an instruction to go read something. It now has a routing row plus a general rule.
- Separately on `tooling/agent-harness-hooks`: the mechanical version, since the routing fix still relies on someone choosing to read. Claude Code never auto-loads `AGENTS.md` at all — a `docs/inflight/CLAUDE.md` importing it now loads automatically the moment anything in that directory is touched.

Comment (astubbs) [reply docs/inflight/test-mockconsumer-harness.md]: Deleted in 3d4737f5 rather than updated — you're right that it didn't earn a file.

What was left after the trim was a warning that branches forked before this one will conflict on these six files, which fails `docs/inflight/AGENTS.md`'s own rule against writing down what a command can answer: git announces a conflict when you hit it, and `gh pr list` answers what astubbs#202 is doing.

The one part no command supplies — keep the harness, re-apply the scenario, rather than re-applying a pre-harness copy of the wiring and silently undoing the extraction — is already in `MockConsumerTestBase`'s class javadoc, and it's next to the code a conflict resolver is actually looking at.

**Review by astubbs — COMMENTED** [review]
lgtm
